### PR TITLE
[MRESOLVER-421] Extend NamedLock API to make possible 1..n resource to lock mapping

### PR DESCRIPTION
**Pull request by cstamas**
This PR implements MRESOLVER-421, basically allowing not only 1:1 resource to lock mapping as Named Locks were originally incepted. Also, change lock name to dedicated "key" type, and this improves tremendously logging and diagnostic (ie. if you used hashing name mapper, good luck to tell what artifact it was).

This PR also adds tests to Redisson as there were no tests in the module since its inception.

---

https://issues.apache.org/jira/browse/MRESOLVER-421

### CODE REVIEW

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java [hunk 1]
```diff
@@ -201,82 +201,70 @@ public void acquire(Collection<? extends Artifact> artifacts, Collection<? exten
                 return;
             }
 
+            final NamedLock namedLock = namedLockFactory.getLock(keys);
```
Comment (cstamas) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java:204]: close if failed to lock

File: maven-resolver-demos/maven-resolver-demo-snippets/src/test/resources/simplelogger.properties [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+org.slf4j.simpleLogger.defaultLogLevel=info
+
```
Comment (michael-o) [inline maven-resolver-demos/maven-resolver-demo-snippets/src/test/resources/simplelogger.properties:19]: Is that different from default?

File: maven-resolver-named-locks-ipc/pom.xml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  <parent>
+    <groupId>org.apache.maven.resolver</groupId>
+    <artifactId>maven-resolver</artifactId>
+    <version>2.0.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>maven-resolver-named-locks-ipc</artifactId>
+
+  <name>Maven Artifact Resolver Named Locks using IPC</name>
+  <description>A synchronization utility implementation using IPC.</description>
```
Comment (michael-o) [inline maven-resolver-named-locks-ipc/pom.xml:32]: This should really be a separate PR since this is one big already.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java [hunk 4]
```diff
@@ -201,82 +201,65 @@ public void acquire(Collection<? extends Artifact> artifacts, Collection<? exten
...
+                        String timeStr = time + " " + timeUnit;
+                        LOGGER.trace(
+                                "Failed to acquire {} lock for '{}' in {}",
+                                shared ? "read" : "write",
+                                namedLock.name(),
+                                time + " " + timeUnit);
+
+                        illegalStateExceptions.add(new IllegalStateException(
+                                "Attempt " + attempt + ": Could not acquire " + (shared ? "read" : "write")
+                                        + " lock for '" + namedLock.name() + "' in " + timeStr));
```
[the remark this answers is not in the document]
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java:231]: Yes, for the message you are right, but for actual _acquire_ all is same remains. Will fix the message.

[the remark this answers is not in the document]
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java:231]: Changed my mind: in fact _nothing_ changed, as what really happened is that the cycle (for many locks) got simply "pulled out" from here, and this message remained same. And was wrong from the very beginning.

[the remark this answers is not in the document]
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/synccontext/named/NamedLockFactoryAdapter.java]: Completely reworked logging: in 1.x we were logging _keys_ which does not make sane (especially if using some hashing name mapper like hgav).

**Review by michael-o — CHANGES_REQUESTED** [review]
The IPC lock is still wrong here.

**Comment by cstamas** [comment]
IPC moved out, please review

File: maven-resolver-named-locks/src/main/java/org/eclipse/aether/named/NamedLockFactory.java [hunk 5]
```diff
@@ -26,10 +31,23 @@ public interface NamedLockFactory {
...
+    /**
+     * Creates or reuses existing {@link NamedLock}. Returns instance MUST BE treated as "resource", best in
+     * try-with-resource block.
+     *
+     * @param names the lock keys, must not be {@code null} and must not be empty collection.
      * @return named lock instance, never {@code null}.
+     * @since 2.0.0
      */
-    NamedLock getLock(String name);
+    NamedLock getLock(Collection<NamedLockKey> names);
```
Comment (michael-o) [inline maven-resolver-named-locks/src/main/java/org/eclipse/aether/named/NamedLockFactory.java:50]: This is inconsistent. The above method has  arg `key`, but this one has `names`, shouldn't it be `keys`?

File: maven-resolver-named-locks/src/main/java/org/eclipse/aether/named/NamedLockKey.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name);
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() + "{" + "name='" + name + '\'' + ", resources='" + resources + '\'' + '}';
```
Comment (michael-o) [inline maven-resolver-named-locks/src/main/java/org/eclipse/aether/named/NamedLockKey.java:81]: I think collections are enclosed with brackets, I guess the single quotes are redundant or confusing.

File: maven-resolver-named-locks/src/main/java/org/eclipse/aether/named/NamedLockKey.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    public static NamedLockKey of(String name, String... resource) {
```
Comment (michael-o) [inline maven-resolver-named-locks/src/main/java/org/eclipse/aether/named/NamedLockKey.java:84]: resources

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by cstamas** [comment]
@michael-o with love, only for you 0338733318627638792a4863e64de49b120282bf :smile:

**Review by michael-o — APPROVED** [review]
From what I see, I don't have any objection.

**Comment by jira-importer** [comment]
Resolve #1099
